adam/learner/contrastive_learner.py

```diff
@@ -244,69 +251,78 @@ def log_hypotheses(self, log_output_path: Path) -> None:
-
-                if (
-                    pattern_to_graph_match.pattern_node_to_matched_graph_node[
-                        pattern_node
-                    ]
-                    in difference_nodes
-                ):
-                    self._ontology_node_present_in_difference[
-                        concept, pattern_node.property_value
-                    ] += 1
-            # Otherwise, if it's categorical, count the value observed
-            elif isinstance(pattern_node, CategoricalPredicate):
-                self._categorical_values_present[concept, pattern_node.value] += 1
-
-                if (
-                    pattern_to_graph_match.pattern_node_to_matched_graph_node[
-                        pattern_node
-                    ]
-                    in difference_nodes
-                ):
-                    self._categorical_values_present_in_difference[
-                        concept, pattern_node.value
-                    ] += 1
+        for pattern_to_graph_match in pattern_to_graph_matches:
+            if pattern_to_graph_match:
+                for pattern_node in pattern_to_graph_match.matched_pattern:
+                    # If it's a stroke GNN recognition node, count the "recognized object" string observed
+                    if isinstance(pattern_node, StrokeGNNRecognitionPredicate):
+                        self._stroke_gnn_recognition_nodes_present[
+                            concept, pattern_node.recognized_object
+                        ] += 1
+
+                        if (
+                            pattern_to_graph_match.pattern_node_to_matched_graph_node[
+                                pattern_node
+                            ]
+                            in difference_nodes
+                        ):
+                            self._stroke_gnn_recognition_nodes_present_in_difference[
+                                concept, pattern_node.recognized_object
+                            ] += 1
+                    # If it has an ontology node, do X
+                    elif isinstance(pattern_node, IsOntologyNodePredicate):
+                        self._ontology_node_present[
+                            concept, pattern_node.property_value
+                        ] += 1
+
+                        if (
+                            pattern_to_graph_match.pattern_node_to_matched_graph_node[
+                                pattern_node
+                            ]
+                            in difference_nodes
+                        ):
+                            self._ontology_node_present_in_difference[
+                                concept, pattern_node.property_value
+                            ] += 1
+                    # Otherwise, if it's categorical, count the value observed
+                    elif isinstance(pattern_node, CategoricalPredicate):
+                        self._categorical_values_present[concept, pattern_node.value] += 1
```

Comment:
Unrelated fix, but small enough that it seems worth including: Now that categoricals also have features, this should also index by `pattern_node.label`:

```suggestion
                        self._categorical_values_present[concept, pattern_node.label, pattern_node.value] += 1
```

This also requires updating the type annotation for `_categorical_values_present`.